### DISCUSSION TITLE
Override Speed

### DISCUSSION THREAD

**Discussion by luni64** [luni64/TeensyStep#124] [Q&A]
### Answer to an email question:   How to use the override speed function to jog a motor

This simple example shows how to use a pot to jog a motor. The center value of the pot is assumed to be 450. Pot values from 451 to 1024 will rotate the motor clockwise (1024 corresponds to max speed) pot values from 449 down to 0 will rotate the motor in the other direction. 

The example uses https://github.com/dxinteractive/ResponsiveAnalogRead to get smooth changes and a stable set value. 

```c++ 
#include "Arduino.h"
#include "ResponsiveAnalogRead.h"
#include "TeensyStep.h"

constexpr int stpPin1    = 0;
constexpr int dirPin1    = 1;
constexpr int potPin1    = A0;
constexpr int potCenter1 = 450;

RotateControl rc1;
Stepper s1(stpPin1, dirPin1); 

ResponsiveAnalogRead pot1(potPin1, false); // read potentiometers through https://github.com/dxinteractive/ResponsiveAnalogRead
                                           // to get smooth changing and stable set values

void setSpeed(RotateControl& controller, ResponsiveAnalogRead& pot, int potCenter)
{
    if (pot.hasChanged())  // only do something if the pot value changed
    {
        float overrideValue = (pot.getValue() - potCenter) / 512.0f; // transform [0..1024] to approximately [-1.0...+1.0]
        controller.overrideSpeed(overrideValue);                     // override speed. 1-> maxSpeed, 0.5 -> half, -0.5 -> half opposite direction...
    }
}

void setup()
{
    pinMode(LED_BUILTIN, OUTPUT);

    s1.setMaxSpeed(15'000);
    s1.setAcceleration(30'000);

    rc1.rotateAsync(s1);              // start rotation. This would accelerate to MaxSpeed, but we...
    setSpeed(rc1, pot1, potCenter1);  // ...immediately override speed to reflect current pot value
}

elapsedMillis stopwatch;

void loop()
{
    setSpeed(rc1, pot1, potCenter1);
    // setSpeed(rc2, pot2, potCenter2);
    // ...

    if (stopwatch > 200) // print stepper data every 200ms
    {
        stopwatch -= 200;
        Serial.printf("s1 pos: %d \n", s1.getPosition());
        digitalToggleFast(LED_BUILTIN);
    }
}

```

**Comment by Michl59** [comment]
Hi Luni64.

Thanks for this code! In order to understand it better I built a test setup and loaded it onto my Teensy.

The motor spins, however the Pot on A0 doesn't change anything atm.
I added 

pot1.update();

at the start of the set Speed, right before the if, now the Pot controls speed and direction.

Is that the "right" way to do this, or would you suggest a more elegant way?

Cheers

Michael

PS: How would the skript change if I needed to define a "dead space" of the stick? Like say the center is pot value 512 and the stick shouldn't so anything when the pot is between 462 and 562? As the motor would need to stop, Im not sure a simple "if" would be sufficient? I could see the skript asking "if pot value between 462 to 562 override speed to 0, else..." and then your above skript

**Reply by luni64** [reply]
Yes, the call to update() is necessary. I tested the code without the pot (used hard coded values) but Murphy's Law is lurking everywhere :-). Or, the devil doesn't sleep as we say here :-) 

If it still doesn't work I'll go into the basement and fetch a pot...

**Reply by Michl59** [reply]
> Yes, the call to update() is necessary. I tested the code without the pot (used hard coded values) but Murphy's Law is lurking everywhere :-). Or, the devil doesn't sleep as we say here :-) 
> 
> 
> 
> If it still doesn't work I'll go into the basement and fetch a pot...

Or "Der Teufel ist manchmal ein Eichhörnchen" :-P


Yes, with the update it works fine, I tried it before posting my reply, wasn't sure if there was a more compact/elegant solution

As to the deadsone question:
I added another variable and renamed the center, so I have a center low and center high

Currently figuring out how to work the if. Thinking one if to see if it's in that range if the deadzone (Motor doesn't rotate( and if not execute your original skript, i'll keep you posted

One thing crossed my mind:
Ia it possible to define a range of steps for the motor?
I have one that can only rotate a certain amount, but freely within that range, however I cannot add sensors to tell it to stop, so the Teensy needs to count.

Is it possible to define a movement range with TeensyStep?

**Comment by luni64** [comment]
To implement a deadband I'd change the `setSpeed` function to the following: 

```c++
void setSpeed(RotateControl& controller, ResponsiveAnalogRead& pot, int potCenter)
{
    pot.update();
    if (pot.hasChanged())                                                 // only do something if the pot value changed
    {
        float overrideValue      = (pot.getValue() - potCenter) / 512.0f; // transform [0..1024] to approximately [-1.0...+1.0]

        constexpr float deadband = 0.1;                                   // deadband 10%
        if (fabsf(overrideValue) < deadband)                              // return 0 if we are in the deadband
            overrideValue = 0.0f;
        else                                                              // outside we need correct the value to start from 0
            overrideValue -= overrideValue > 0 ? deadband : -deadband;    // depending on the sign we need to add or substract the deadband value to start from 0 when leaving the deadband

        controller.overrideSpeed(overrideValue);
    }
}
```
Regarding a limited range: 
There is no such thing for the rotateControler. But you can do something like this: 
```c++
void loop()
{
    setSpeed(rc1, pot1, potCenter1);
    // setSpeed(rc2, pot2, potCenter2);
    // ...

    int32_t pos = s1.getPosition();
    if (pos > 1000 || pos < -2000)  // some range
    {
        rc1.stop();          // decelerates to 0, make sure to have enough headroom for deceleration
     // rc1.emergencyStop(); // hard stop, motor will probably lose steps
    }

    if (stopwatch > 200) // print stepper data every 200ms
    {
        stopwatch -= 200;
        Serial.printf("s1 pos: %d \n", s1.getPosition());
        digitalToggleFast(LED_BUILTIN);
    }
}
```

**Reply by Michl59** [reply]
Wicked! I'll tinker with it tomorrow :)

Cheers
